File: src/boj/랜선자르기_1654.py
import sys

input = sys.stdin.readline

K, N = map(int, input().split())
arr = []
left = 1
right = -1
ans = -1

for _ in range(K):
    n = int(input())
    arr.append(n)
    right = max(right, n)

while left <= right:
    mid = (left + right) // 2

    cnt = 0
    for n in arr:
        cnt += (n // mid)

    if cnt >= N:
        left = mid + 1
        ans = max(ans, mid)
    else:
        right = mid - 1

print(ans)

File: src/boj/랜선자르기_1654.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class 랜선자르기_1654 {
	
	static int[] arr;
	static int K, N;
	static long ans;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		K = Integer.parseInt(st.nextToken());    // 10,000 이하의 정수 (랜선 K개)
		N = Integer.parseInt(st.nextToken());	// 1,000,000 이하의 정수 (필요한 랜선 N개)
		arr = new int[K];

		for (int i = 0; i < K; i++) arr[i] = Integer.parseInt(br.readLine());
		Arrays.sort(arr);

		long left = 1, right = arr[K - 1];
		long mid;
		while (left <= right) {
			mid = (left + right) / 2;

			if (test(mid)) {
				left = mid + 1;
			} else {
				right = mid - 1;
			}
		}

		System.out.println(ans);
	}

	static boolean test(long val) {
		int sum = 0;
		for (int i = 0; i < K; i++) sum += arr[i] / val;

		if (sum >= N) {
			if (ans < val) ans = val;
			return true;
		}

		return false;
	}
}
